File: meme_generator.js
// meme_generator.js
#!/usr/bin/env node
'use strict';

const { createCanvas, loadImage, registerFont } = require('canvas');
const fs = require('fs');
const path = require('path');

function parseHexColor(hex) {
    if (hex.length === 7 && hex[0] === '#') {
        const r = parseInt(hex.substring(1, 3), 16);
        const g = parseInt(hex.substring(3, 5), 16);
        const b = parseInt(hex.substring(5, 7), 16);
        return `rgb(${r},${g},${b})`;
    }
    return 'rgb(255,255,255)';
}

async function memeGenerator(inputPath, outputPath, topText, bottomText,
                            fontSize, fontPath, color, strokeColor, strokeWidth, shadow, padding) {
    const img = await loadImage(inputPath);
    const w = img.width, h = img.height;
    const canvas = createCanvas(w, h);
    const ctx = canvas.getContext('2d');
    ctx.drawImage(img, 0, 0);

    if (fontPath) {
        registerFont(fontPath, { family: 'Custom' });
        ctx.font = `bold ${fontSize}px Custom`;
    } else {
        ctx.font = `bold ${fontSize}px Impact, sans-serif`;
    }
    ctx.textAlign = 'center';
    ctx.textBaseline = 'top';
    ctx.fillStyle = parseHexColor(color);
    ctx.strokeStyle = parseHexColor(strokeColor);
    ctx.lineWidth = strokeWidth;

    // Верхний текст
    if (topText) {
        const x = w / 2;
        const y = padding;
        ctx.strokeText(topText, x, y);
        ctx.fillText(topText, x, y);
    }

    // Нижний текст
    if (bottomText) {
        ctx.textBaseline = 'bottom';
        const x = w / 2;
        const y = h - padding;
        ctx.strokeText(bottomText, x, y);
        ctx.fillText(bottomText, x, y);
    }

    const out = fs.createWriteStream(outputPath);
    const stream = canvas.createPNGStream();
    stream.pipe(out);
    return new Promise((resolve, reject) => {
        out.on('finish', resolve);
        out.on('error', reject);
    });
}

async function batchProcess(inputDir, outputDir, opts) {
    if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });
    const exts = ['.png', '.jpg', '.jpeg', '.bmp', '.tiff', '.webp'];
    const files = fs.readdirSync(inputDir).filter(f => exts.includes(path.extname(f).toLowerCase()));
    for (let i = 0; i < files.length; i++) {
        const inPath = path.join(inputDir, files[i]);
        const outPath = path.join(outputDir, files[i]);
        console.log(`[${i+1}/${files.length}] ${files[i]}`);
        await memeGenerator(inPath, outPath, opts.top, opts.bottom, opts.fontSize,
                            opts.fontPath, opts.color, opts.stroke, opts.strokeWidth,
                            opts.shadow, opts.padding);
    }
}

async function main() {
    const args = process.argv.slice(2);
    let input, output = 'meme.png', top = '', bottom = '';
    let fontSize = 48, fontPath = null, color = '#FFFFFF', stroke = '#000000';
    let strokeWidth = 2, shadow = false, padding = 20, batchDir = null;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '-t' && i+1 < args.length) top = args[++i];
        else if (args[i] === '-b' && i+1 < args.length) bottom = args[++i];
        else if (args[i] === '-o' && i+1 < args.length) output = args[++i];
        else if (args[i] === '-s' && i+1 < args.length) fontSize = parseInt(args[++i], 10);
        else if (args[i] === '-c' && i+1 < args.length) color = args[++i];
        else if (args[i] === '-f' && i+1 < args.length) fontPath = args[++i];
        else if (args[i] === '--stroke' && i+1 < args.length) stroke = args[++i];
        else if (args[i] === '--stroke-width' && i+1 < args.length) strokeWidth = parseInt(args[++i], 10);
        else if (args[i] === '--shadow') shadow = true;
        else if (args[i] === '--padding' && i+1 < args.length) padding = parseInt(args[++i], 10);
        else if (args[i] === '--batch' && i+1 < args.length) batchDir = args[++i];
        else if (args[i] === '-h' || args[i] === '--help') {
            console.log(`Usage: node meme_generator.js <input> [options]
  -t <text>    Top text
  -b <text>    Bottom text
  -o <file>    Output
  -s <N>       Font size
  -c <hex>     Color
  -f <file>    Font path
  --stroke <hex>  Stroke color
  --stroke-width <N>  Stroke width
  --shadow     Add shadow
  --padding <N>  Padding
  --batch <dir>  Batch`);
            process.exit(0);
        } else if (!input) input = args[i];
    }
    if (!input) { console.error('Укажите входной файл или папку.'); process.exit(1); }

    const opts = { top, bottom, fontSize, fontPath, color, stroke, strokeWidth, shadow, padding };

    if (batchDir) {
        await batchProcess(input, batchDir, opts);
        console.log(`✅ Batch processing completed. Results in ${batchDir}`);
    } else {
        await memeGenerator(input, output, top, bottom, fontSize, fontPath, color, stroke, strokeWidth, shadow, padding);
        console.log(`✅ Meme saved to ${output}`);
    }
}

main().catch(console.error);

File: meme_generator.rb
#!/usr/bin/env ruby
# meme_generator.rb
# encoding: UTF-8

require 'rmagick'
include Magick
require 'optparse'
require 'fileutils'

def parse_hex_color(hex)
  if hex =~ /^#([0-9a-f]{6})$/i
    rgb = hex[1..-1].scan(/../).map { |c| c.to_i(16) }
    return "rgb(#{rgb[0]},#{rgb[1]},#{rgb[2]})"
  end
  'rgb(255,255,255)'
end

def meme_generator(input_path, output_path, top_text, bottom_text,
                   font_size, font_path, color, stroke_color, stroke_width, shadow, padding)
  img = Image.read(input_path).first
  w, h = img.columns, img.rows

  draw = Draw.new
  draw.font = font_path if font_path && File.exist?(font_path)
  draw.pointsize = font_size
  draw.fill = parse_hex_color(color)
  draw.stroke = parse_hex_color(stroke_color)
  draw.stroke_width = stroke_width
  draw.gravity = NorthGravity
  draw.annotate(img, 0, 0, 0, padding, top_text) if top_text && !top_text.empty?

  draw.gravity = SouthGravity
  draw.annotate(img, 0, 0, 0, padding, bottom_text) if bottom_text && !bottom_text.empty?

  img.write(output_path)
end

def batch_process(input_dir, output_dir, opts)
  FileUtils.mkdir_p(output_dir)
  exts = %w[.png .jpg .jpeg .bmp .tiff .webp]
  files = Dir.entries(input_dir)
             .select { |f| exts.include?(File.extname(f).downcase) }
             .map { |f| File.join(input_dir, f) }
  files.each_with_index do |f, i|
    out_path = File.join(output_dir, File.basename(f))
    puts "[#{i+1}/#{files.size}] #{File.basename(f)}"
    meme_generator(f, out_path, opts[:top], opts[:bottom], opts[:font_size],
                   opts[:font_path], opts[:color], opts[:stroke], opts[:stroke_width],
                   opts[:shadow], opts[:padding])
  end
end

def main
  options = { top: '', bottom: '', output: 'meme.jpg', font_size: 48,
              font_path: nil, color: '#FFFFFF', stroke: '#000000',
              stroke_width: 2, shadow: false, padding: 20, batch: nil }
  input = nil

  OptionParser.new do |opts|
    opts.banner = "Usage: meme_generator.rb <input> [options]"
    opts.on('-t TEXT', 'Top text') { |v| options[:top] = v }
    opts.on('-b TEXT', 'Bottom text') { |v| options[:bottom] = v }
    opts.on('-o FILE', 'Output') { |v| options[:output] = v }
    opts.on('-s N', Integer, 'Font size') { |v| options[:font_size] = v }
    opts.on('-c HEX', 'Color') { |v| options[:color] = v }
    opts.on('-f FILE', 'Font path') { |v| options[:font_path] = v }
    opts.on('--stroke HEX', 'Stroke color') { |v| options[:stroke] = v }
    opts.on('--stroke-width N', Integer, 'Stroke width') { |v| options[:stroke_width] = v }
    opts.on('--shadow', 'Add shadow') { options[:shadow] = true }
    opts.on('--padding N', Integer, 'Padding') { |v| options[:padding] = v }
    opts.on('--batch DIR', 'Batch output dir') { |v| options[:batch] = v }
    opts.on('-h', 'Help') { puts opts; exit }
  end.parse!

  input = ARGV[0] if ARGV[0]
  unless input
    puts "Укажите входной файл или папку."
    exit 1
  end

  if options[:batch]
    batch_process(input, options[:batch], options)
    puts "✅ Batch processing completed. Results in #{options[:batch]}"
  else
    meme_generator(input, options[:output], options[:top], options[:bottom],
                   options[:font_size], options[:font_path], options[:color],
                   options[:stroke], options[:stroke_width], options[:shadow],
                   options[:padding])
    puts "✅ Meme saved to #{options[:output]}"
  end
end

main if __FILE__ == $0
